x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java

```diff
@@ -0,0 +1,434 @@
+            throw new ElasticsearchSecurityException("Failed to consume the OpenID connect response");
+        }
+    }
+
+    /**
+     * Validates an access token according to the
+     * <a href="https://openid.net/specs/openid-connect-core-1_0.html#ImplicitTokenValidation">specification</a>
+     *
+     * @param accessToken The Access Token to validate
+     * @param idToken     The Id Token that was received in the same response
+     * @param optional    When using the authorization code flow the OP might not provide the at_hash parameter in the
+     *                    Id Token as allowed in the specification. In such a case we can't validate the access token
+     *                    but this is considered safe as it was received in a back channel communication that was protected
+     *                    by TLS.
+     */
+    private void validateAccessToken(AccessToken accessToken, JWT idToken, boolean optional) {
+        try {
+            // only "bearer" is defined in the specification but check just in case
+            if (accessToken.getType().toString().equals("Bearer") == false) {
+                logger.debug("Invalid access token type [{}], while [Bearer] was expected", accessToken.getType());
+                throw new ElasticsearchSecurityException("Received a response with an invalid access token type");
+            }
+            String atHashValue = idToken.getJWTClaimsSet().getStringClaim("at_hash");
+            if (null == atHashValue && optional == false) {
+                logger.debug("Failed to verify access token. at_hash claim is missing from the ID Token");
+                throw new ElasticsearchSecurityException("Failed to verify access token");
+            }
+            AccessTokenHash atHash = new AccessTokenHash(atHashValue);
+            JWSAlgorithm jwsAlgorithm = JWSAlgorithm.parse(idToken.getHeader().getAlgorithm().getName());
+            AccessTokenValidator.validate(accessToken, jwsAlgorithm, atHash);
+        } catch (Exception e) {
+            logger.debug("Failed to verify access token.", e);
+            throw new ElasticsearchSecurityException("Failed to verify access token.");
+        }
+    }
+
+    /**
+     * Parses and validates an OpenID Connect Id Token to a set of claims
+     *
+     * @param idToken       The Id Token to parse and validate
+     * @param expectedNonce The nonce that was generated in the beginning of this authentication attempt and was stored at the user's
+     *                      session with the facilitator
+     * @return a {@link JWTClaimsSet} with the OP claims that were contained in the Id Token
+     */
+    @SuppressForbidden(reason = "uses toFile")
+    private JWTClaimsSet validateAndParseIdToken(JWT idToken, Nonce expectedNonce) {
+        Secret clientSecret = null;
+        try {
+            final IDTokenValidator validator;
+            final JWSAlgorithm requestedAlgorithm = rpConfig.getSignatureVerificationAlgorithm();
```

Comment:
Probably a matter of taste, but I would move the `signatureVerificationAlgorithm` from the RP config to the OP config and rename it `signatureAlgorithm`.

Reply:
This is actually an RP setting. The relying party dictates the signature algorithm it needs the OP to sign the JWTs with and it validates it upon receipt. I'll change it to `signatureAlgorithm`